### PR TITLE
Fix autodiff bug and add debug helpers for optimizer

### PR DESCRIPTION
**Pull request by gordoncaleb**
This PR fixes a bug with tensor autodiff. It also adds a `onGradientCalculation(...)` and a `onFitnessCalculation(...)` for easy debug of optimizer convergence checking.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/tensor/TensorFitnessFunctionWithGradient.java [hunk 1]
```diff
@@ -52,7 +77,11 @@ public MultivariateFunction fitness() {
         for (Vertex<DoubleTensor> vertex : latentVertices) {
             DoubleTensor tensor = diffs.get(vertex.getId());
             if (tensor != null) {
-                tensors.add(tensor);
+                if (TensorShape.isScalar(vertex.getShape()) && !TensorShape.isScalar(tensor.getShape())) {
```
Comment (joncoello83) [inline keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/tensor/TensorFitnessFunctionWithGradient.java]: This is quite nested and could potentially be made more expressive if this were extracted into a separate function.

File: keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java [hunk 2]
```diff
@@ -47,11 +54,15 @@ public void linearRegression1FactorVariationalMAP() {
 
         double expectedM = 3.0;
         double expectedB = 20.0;
-        List<Point> points = make1FactorTestData(1000, expectedM, expectedB);
+//        List<Point> points = make1FactorTestData(10000, expectedM, expectedB);
```
Comment (joncoello83) [inline keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java]: Commented code

Comment (gordoncaleb) [reply keanu-project/src/test/java/io/improbable/keanu/e2e/regression/LinearRegression.java]: Good catch!

File: keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/tensor/TensorFitnessFunctionWithGradient.java (hunk 1 above)
Comment (gordoncaleb) [reply keanu-project/src/main/java/io/improbable/keanu/algorithms/variational/tensor/TensorFitnessFunctionWithGradient.java]: Yup, agreed.

**Pull request merged**